dhcpv4/ztpv4/ztp.go

```diff
@@ -15,62 +17,107 @@ type VendorData struct {
+			for _, f := range bytes.Split(id.Data, []byte(";")) {
+				p := bytes.Split(f, []byte(":"))
+				if len(p) != 2 {
+					return nil, errVendorOptionMalformed
+				}
 
-	switch {
-	// Arista;DCS-7050S-64;01.23;JPE12221671
-	case strings.HasPrefix(vc, "Arista;"):
-		p := strings.Split(vc, ";")
-		if len(p) < 4 {
-			return nil, errVendorOptionMalformed
+				switch string(p[0]) {
+				case "SN":
+					vd.Serial = string(p[1])
+				case "PID":
+					vd.Model = string(p[1])
+				}
+			}
 		}
+	}
+	return vd, nil
+}
 
-		vd.VendorName = p[0]
-		vd.Model = p[1]
-		vd.Serial = p[3]
-		return vd, nil
+func parseClassIdentifier(packet *dhcpv4.DHCPv4) (*VendorData, error) {
+	vd := &VendorData{}
+	vc := packet.ClassIdentifier()
 
-	// ZPESystems:NSC:002251623
-	case strings.HasPrefix(vc, "ZPESystems:"):
-		p := strings.Split(vc, ":")
-		if len(p) < 3 {
-			return nil, errVendorOptionMalformed
-		}
+	switch {
+		// Arista;DCS-7050S-64;01.23;JPE12221671
+		case strings.HasPrefix(vc, "Arista;"):
+			p := strings.Split(vc, ";")
+			if len(p) < 4 {
+				return nil, errVendorOptionMalformed
+			}
 
-		vd.VendorName = p[0]
-		vd.Model = p[1]
-		vd.Serial = p[2]
-		return vd, nil
+			vd.VendorName = p[0]
+			vd.Model = p[1]
+			vd.Serial = p[3]
+			return vd, nil
+
+		// ZPESystems:NSC:002251623
+		case strings.HasPrefix(vc, "ZPESystems:"):
+			p := strings.Split(vc, ":")
+			if len(p) < 3 {
+				return nil, errVendorOptionMalformed
+			}
 
-	// Juniper option 60 parsing is a bit more nuanced.  The following are all
-	// "valid" identifying stings for Juniper:
-	//    Juniper-ptx1000-DD576      <vendor>-<model>-<serial
-	//    Juniper-qfx10008           <vendor>-<model> (serial in hostname option)
-	//    Juniper-qfx10002-361-DN817 <vendor>-<model>-<serial> (model has a dash in it!)
-	case strings.HasPrefix(vc, "Juniper-"):
-		p := strings.Split(vc, "-")
-		if len(p) < 3 {
+			vd.VendorName = p[0]
 			vd.Model = p[1]
-			vd.Serial = packet.HostName()
-			if len(vd.Serial) == 0 {
-				return nil, errors.New("host name option is missing")
+			vd.Serial = p[2]
+			return vd, nil
+
+		// Juniper option 60 parsing is a bit more nuanced.  The following are all
+		// "valid" identifying stings for Juniper:
+		//    Juniper-ptx1000-DD576      <vendor>-<model>-<serial
+		//    Juniper-qfx10008           <vendor>-<model> (serial in hostname option)
+		//    Juniper-qfx10002-361-DN817 <vendor>-<model>-<serial> (model has a dash in it!)
+		case strings.HasPrefix(vc, "Juniper-"):
+			p := strings.Split(vc, "-")
+			if len(p) < 3 {
+				vd.Model = p[1]
+				vd.Serial = packet.HostName()
+				if len(vd.Serial) == 0 {
+					return nil, errors.New("host name option is missing")
+				}
+			} else {
+				vd.Model = strings.Join(p[1:len(p)-1], "-")
+				vd.Serial = p[len(p)-1]
 			}
-		} else {
-			vd.Model = strings.Join(p[1:len(p)-1], "-")
-			vd.Serial = p[len(p)-1]
+
+			vd.VendorName = p[0]
+			return vd, nil
 		}
 
-		vd.VendorName = p[0]
+	return vd, nil
```

Comment:
Should it return nil `VendorData` here and check the pointer is not null afterwards instead of checking for an empty struct?

Reply:
I will return a nil, nil when VendorData is not set in either functions and return if it is not nil in ParseVendorData